File: apps/ios/Pathfinding/Pathfinding/Models/PdfExport.swift
import Foundation
import SwiftUI

// MARK: - PDF Export Types

/// PDF template types
enum PdfTemplateType: String, Codable, CaseIterable {
  case classic
  case modern
  case minimal
  case colorful

  var displayName: String {
    switch self {
    case .classic: return "经典"
    case .modern: return "现代"
    case .minimal: return "极简"
    case .colorful: return "多彩"
    }
  }

  var description: String {
    switch self {
    case .classic: return "传统衬线字体，优雅边框设计"
    case .modern: return "简洁无衬线设计，柔和阴影效果"
    case .minimal: return "超简洁设计，极少装饰元素"
    case .colorful: return "鲜艳渐变色，活力四射"
    }
  }

  var previewColor: String {
    switch self {
    case .classic: return "#8B4513"
    case .modern: return "#6366F1"
    case .minimal: return "#374151"
    case .colorful: return "#EC4899"
    }
  }

  /// SwiftUI Color for preview
  var color: Color {
    switch self {
    case .classic: return .brown
    case .modern: return .indigo
    case .minimal: return .gray
    case .colorful: return .pink
    }
  }
}

/// PDF language options
enum PdfLanguage: String, Codable, CaseIterable {
  case zh
  case en
  case bilingual

  var displayName: String {
    switch self {
    case .zh: return "中文"
    case .en: return "English"
    case .bilingual: return "双语"
    }
  }
}

/// PDF page size options
enum PdfPageSize: String, Codable, CaseIterable {
  case A4
  case Letter
  case A5

  var displayName: String {
    switch self {
    case .A4: return "A4"
    case .Letter: return "Letter"
    case .A5: return "A5"
    }
  }

  var description: String {
    switch self {
    case .A4: return "210 x 297 mm"
    case .Letter: return "8.5 x 11 in"
    case .A5: return "148 x 210 mm"
    }
  }
}

// MARK: - PDF Export Options

/// Options for PDF export
struct PdfExportOptions: Codable {
  var template: String?
  var language: String?
  var pageSize: String?
  var includeMap: Bool?
  var includeCover: Bool?
  var includeToc: Bool?
  var includePhotos: Bool?
  var includeTransport: Bool?
  var primaryColor: String?
  var authorName: String?

  enum CodingKeys: String, CodingKey {
    case template
    case language
    case pageSize = "page_size"
    case includeMap = "include_map"
    case includeCover = "include_cover"
    case includeToc = "include_toc"
    case includePhotos = "include_photos"
    case includeTransport = "include_transport"
    case primaryColor = "primary_color"
    case authorName = "author_name"
  }

  /// Create default options
  static var `default`: PdfExportOptions {
    PdfExportOptions(
      template: PdfTemplateType.modern.rawValue,
      language: PdfLanguage.zh.rawValue,
      pageSize: PdfPageSize.A4.rawValue,
      includeMap: true,
      includeCover: true,
      includeToc: true,
      includePhotos: true,
      includeTransport: true,
      primaryColor: "#6366F1",
      authorName: nil
    )
  }
}

File: apps/api/src/routes/pdf-export.ts
/**
 * PDF Export API Routes
 * Endpoints for generating PDF itinerary documents
 */

import { zValidator } from '@hono/zod-validator';
import { Hono } from 'hono';
import { z } from 'zod';
import { PdfExportService } from '../services/pdfExportService';

interface Variables {
  userId: string;
  accessToken: string;
}

// Validation schemas
const PdfOptionsSchema = z.object({
  template: z.enum(['classic', 'modern', 'minimal', 'colorful']).optional(),
  language: z.enum(['zh', 'en', 'bilingual']).optional(),
  page_size: z.enum(['A4', 'Letter', 'A5']).optional(),
  include_map: z.boolean().optional(),
  include_cover: z.boolean().optional(),
  include_toc: z.boolean().optional(),
  include_photos: z.boolean().optional(),
  include_transport: z.boolean().optional(),
  primary_color: z
    .string()
    .regex(/^#[0-9A-Fa-f]{6}$/, 'Invalid hex color')
    .optional(),
  author_name: z.string().max(100).optional(),
});

// Public routes (no auth required)
export const publicPdfExportRoutes = new Hono();

/**
 * GET /pdf/templates - Get available PDF templates
 * No authentication required
 */
publicPdfExportRoutes.get(
  '/templates',
  zValidator(
    'query',
    z.object({
      lang: z.enum(['zh', 'en']).optional().default('zh'),
    })
  ),
  async (c) => {
    const { lang } = c.req.valid('query');

    const templates = await PdfExportService.getTemplates(lang);

    return c.json({
      success: true,
      data: templates,
    });
  }
);

// Protected routes (auth required)
export const pdfExportRoutes = new Hono<{ Variables: Variables }>();

/**
 * GET /pdf/itineraries/:id/preview - Get PDF preview info for an itinerary
 */
pdfExportRoutes.get('/itineraries/:id/preview', async (c) => {
  const userId = c.get('userId');
  const accessToken = c.get('accessToken');
  const itineraryId = c.req.param('id');

  const preview = await PdfExportService.getPreview(
    itineraryId,
    userId,
    accessToken
  );

  return c.json({
    success: true,
    data: preview,
  });
});

/**
 * POST /pdf/itineraries/:id - Generate PDF for an itinerary
 * Returns the PDF file as binary response
 */
pdfExportRoutes.post(
  '/itineraries/:id',
  zValidator('json', PdfOptionsSchema),
  async (c) => {
    const userId = c.get('userId');
    const accessToken = c.get('accessToken');
    const itineraryId = c.req.param('id');
    const options = c.req.valid('json');

    const result = await PdfExportService.generatePdf(
      itineraryId,
      userId,
      {
        template: options.template,
        language: options.language,
        pageSize: options.page_size,
        includeMap: options.include_map,
        includeCover: options.include_cover,
        includeToc: options.include_toc,
        includePhotos: options.include_photos,
        includeTransport: options.include_transport,
        primaryColor: options.primary_color,
        authorName: options.author_name,
      },
      accessToken
    );

    return new Response(result.data, {
      status: 200,
      headers: {
        'Content-Type': result.contentType,
        'Content-Disposition': `attachment; filename="${encodeURIComponent(result.filename)}"`,
        'Content-Length': result.data.byteLength.toString(),
        'X-Page-Count': result.pageCount.toString(),
      },
    });
  }
);

/**
 * POST /pdf/guides/:id - Generate PDF for a travel guide
 * Proxies to crawler service's guide PDF endpoint
 */
pdfExportRoutes.post(
  '/guides/:id',
  zValidator('json', PdfOptionsSchema),
  async (c) => {
    const guideId = c.req.param('id');
    const options = c.req.valid('json');

    const result = await PdfExportService.generateGuidePdf(guideId, {
      template: options.template,
      language: options.language,
      pageSize: options.page_size,
      includeMap: options.include_map,
      includeCover: options.include_cover,
      includeToc: options.include_toc,
      includePhotos: options.include_photos,
      includeTransport: options.include_transport,
      primaryColor: options.primary_color,
      authorName: options.author_name,
    });

    return new Response(result.data, {
      status: 200,
      headers: {
        'Content-Type': result.contentType,
        'Content-Disposition': `attachment; filename="${encodeURIComponent(result.filename)}"`,
        'Content-Length': result.data.byteLength.toString(),
        'X-Page-Count': result.pageCount.toString(),
      },
    });
  }
);
